server/src/main/java/org/elasticsearch/action/admin/indices/template/post/TransportSimulateIndexTemplateAction.java

```diff
@@ -0,0 +1,132 @@
+                                                ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver,
+                                                NamedXContentRegistry xContentRegistry) {
+        super(SimulateIndexTemplateAction.NAME, transportService, clusterService, threadPool, actionFilters,
+            SimulateIndexTemplateRequest::new, indexNameExpressionResolver);
+        this.indexTemplateService = indexTemplateService;
+        this.xContentRegistry = xContentRegistry;
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected SimulateIndexTemplateResponse read(StreamInput in) throws IOException {
+        return new SimulateIndexTemplateResponse(in);
+    }
+
+    @Override
+    protected void masterOperation(Task task, SimulateIndexTemplateRequest request, ClusterState state,
+                                   ActionListener<SimulateIndexTemplateResponse> listener) throws Exception {
+        ClusterState simulateOnClusterState = state;
+        if (request.getIndexTemplateRequest() != null) {
+            // we'll "locally" add the template defined by the user in the cluster state (as if it existed in the system)
+            String simulateTemplateToAdd = "simulate_new_template_" + UUIDs.randomBase64UUID();
+            simulateOnClusterState = indexTemplateService.addIndexTemplateV2(state, request.getIndexTemplateRequest().create(),
+                simulateTemplateToAdd, request.getIndexTemplateRequest().indexTemplate());
+        }
+
+        String matchingTemplate = findV2Template(simulateOnClusterState.metadata(), request.getIndexName(), false);
+        if (matchingTemplate == null) {
+            listener.onResponse(new SimulateIndexTemplateResponse(null, null));
+            return;
+        }
+        Settings settings = resolveSettings(simulateOnClusterState.metadata(), matchingTemplate);
+
+        // empty request mapping as the user can't specify any explicit mappings via the simulate api
+        Map<String, Object> mappings = resolveV2Mappings("{}", simulateOnClusterState, matchingTemplate, xContentRegistry);
+        String mappingsJson = Strings.toString(XContentFactory.jsonBuilder()
+            .startObject()
+            .field(MapperService.SINGLE_MAPPING_NAME, mappings)
+            .endObject());
+
+        List<Map<String, AliasMetadata>> resolvedAliases = MetadataIndexTemplateService.resolveAliases(simulateOnClusterState.metadata(),
+            matchingTemplate);
+        Map<String, AliasMetadata> aliases = new HashMap<>();
+        for (Map<String, AliasMetadata> alias : resolvedAliases) {
+            for (Map.Entry<String, AliasMetadata> aliasEntry : alias.entrySet()) {
+                if (aliases.containsKey(aliasEntry.getKey()) == false) {
+                    aliases.put(aliasEntry.getKey(), aliasEntry.getValue());
+                }
+            }
+        }
```

Comment:
I think we might want to put this logic in a static method in `MetadataCreateIndexService` (since I believe it does something similar) so that the simulation can reuse the real method.